### DISCUSSION TITLE
Identifying pattern in Audio File

### DISCUSSION THREAD

**Discussion by aqibmumtaz** [boris-kz/CogAlg#44] [Ideas]
How to find Pattern in an Audio format:

These are two approaches:

1. Directly convert the audio file into byte ndarray and apply simple all the 1-D(line pattern and line pps) algorithms we have created
Each sample is stored as a number (two bytes) 
range of available combinations:
– 16 bits, 216 = 65,536 
–we want both positive and negative values (To indicate compressions and rarefactions)
 –we use one bit to indicate positive (0) or negative (1)? 
– That leaves us with 15 bits – 15 bits, 215 = 32,768 
– One of those combinations will stand for zero ( We’ll use a “positive” one, so that’s one less pattern for positives so the range is from -32,768 to 32,767)
A Sound has many values in it 
– Numbers that represent the sound at that time in the sample
 • We can get an array of SoundSample objects
 – SoundSample[] sampleArray = sound1.getSamples();
Once we have array we can implement line-pattern and line PPs algos on these arrays?(what do you say)

 2. Calculate the DFT from FFT of the audio signal and store that in the array we can proceed on that array too
OR
further from DFT apply windows and create and plot Spectrogram(In a spectrogram representation plot — one axis represents the time, the second axis represents frequencies and the colors represent magnitude (amplitude) of the observed frequency at a particular time.) and this will convert it into a image like we have now then we can find patterns or line of patterns as we are on 1-D line pattern and 2-D frame pattern identification thing (your thoughts)?

**Comment by kwcckw** [comment]
interesting, so how would be the pattern in audio file looks like?

For images, we may have low level feature, such as lines or edges, and then to blobs and shape, and with that we may achieve some higher level application such as recognition or detection.

In this case, for audio, would it be some applications such as background noise removal? Or auto tuning? To enable the visualization, what would be some other option to 'see' the frequency response other than spectrogram ?

And do you mind to share the code as well?

**Reply by aqibmumtaz** [reply]
Images are 2-D matrix and have pixels like at 1,1 and 1,2 coordinates etc, for audio signals as they are 1-D while we have for every sampling rate audio at every time point/instant we have amplitude value as a low level feature.

As far as I know FFT(Fast Fourier Transform) is the methods to plot frequency response of audio signal. Its an algorithm for the implementation of DFT(Discrete Fourier Transform)

**Comment by boris-kz** [comment]
Not sure DFT is directly applicable here, let's do without it first. Maybe the alg will rediscover it.
As I mentioned before, time is the only dimension here, intensity and frequency are two parameters of input, not dimensions.
Those parameters are all positive, sign should only be in differences and deviations computed by cross-comp.
So, line_patterns will have to cross-compare two parameters in parallel, and initial pattern-defining sign will be a deviation of intensity match + frequency match? But in this case, intensity does represent physical force applied to generate sound, so the match should be defined directly, as min(i1,i2)? This is not the case with frequency though, it's match should be defined indirectly, as ave - abs(diff).

**Reply by kwcckw** [reply]
> So, line_patterns will have to cross-compare two parameters in parallel, and initial pattern-defining sign will be a deviation of intensity match + frequency match? 

Deviation of intensity match + frequency match is the difference between both matches （intensity match  - frequency match) ?

**Reply by boris-kz** [reply]
Sorry, this will be a sum of deviations: (min( i1, i2 ) - ave) + (ave - abs( diff ( f1, f2 ))).
Also, we will need a coeff filter to to represent relative value of these two component matches. 
The problem with that, we will have multiple frequencies at the same time. So that f would have to be an average, which is not very meaningful. The alternative is forming separate patterns for each frequency and then combining them. But that's assuming we have a fixed discrete frequency spectrum, which is not likely.

**Reply by kwcckw** [reply]
> 
> 
> Sorry, this will be a sum of deviations: (min( i1, i2 ) - ave) + (ave - abs( diff ( f1, f2 ))).
> Also, we will need a coeff filter to to represent relative value of these two component matches.

so the match of :
amplitude = (min( i1, i2 ) - ave)
frequency = (ave - abs( diff ( f1, f2 )))

the purpose of coeff filter is to map these 2 parameters into a common measure-able dimension right?

So the sum of deviations would be:
((min( i1, i2 ) - ave)*coef_amplitude) + ((ave - abs( diff ( f1, f2 )))*coef_frequency) ?

But suppose we need to get their relationship first? How would be the amplitude related to frequency , time and vice-versa, then from there we deduce the coef?

**Reply by boris-kz** [reply]
Coef is filter, just like all other filters it's initialized as a guess, then adjusted by feedback. But again, it might not be very meaningful to use combined frequency per "instant" (time pixel) as a comparand...

**Reply by boris-kz** [reply]
Fourie transform would give us a spectrum, so we could form amplitude / time patterns in each component of that spectrum. But that's assuming there is fixed spectrum, vs. simply drifting mix of frequencies. I suppose we can test it on music or specific voice, where the spectrum is well-defined. But the whole thing is artificial...

**Reply by kwcckw** [reply]
I see, yea suppose it would be more efficient to work in those data in frequency domain , instead of time domain. For the defined spectrum, why wouldn't we just use the human audible spectrum? With that we can actually hear the output too.

**Reply by boris-kz** [reply]
Frequency domain here would be a set of recurring / close frequencies, and temporal patterns will be formed for each of them. The next step will be cross-comp among temporally adjacent patterns of different frequencies, similar to comp_slice_. 
Yes, we will use the human audible spectrum, but we don't know what that set of recurring frequencies is, within the spectrum.
Did you see my last top comment?

**Reply by kwcckw** [reply]
Yes, but i'm not familiar with the recurring frequency, let me check on it first, thanks.

**Reply by aqibmumtaz** [reply]
the intensity you mean is amplitude of the signal at one time instance?

**Reply by aqibmumtaz** [reply]
"The problem with that, we will have multiple frequencies at the same time."

it means when we convert time to frequency domain for one time instant we have multiple frequencies?

**Reply by aqibmumtaz** [reply]
when you are talking about having Time as a dimension and taking frequency and intensity as inputs 
you are talking about time-frequency overlapped signals?

**Reply by boris-kz** [reply]
@aqibmumtaz, yes intensity is amplitude here, it means the same thing, I want to use common terms for both images and audio.  

> it means when we convert time to frequency domain for one time instant we have multiple frequencies?

Yes, is that how it works in DFT? I need to get into details of FT, does it work like above-minimal-amplitude cross-comparison at incremental distance, as I proposed in the last comment?

> you are talking about time-frequency overlapped signals?

I mistakenly assumed that frequency is a parameter of input in raw audio, kind of like colors in images. It turned out that we have to compute component frequencies via some kind FT first. 

Then we form time-domain patterns of intensity for each component frequency, with current line_patterns. And then we cross-compare time-overlapped patterns between frequencies, similar to comp_slice_ module in 2D alg. So, time domain will be a primary dimension, in which we define Ps, and frequency domain is secondary, in which we define time-frequency overlapped PPs. 

Does it make sense?

**Reply by kwcckw** [reply]
> 
> 
> Frequency domain here would be a set of recurring / close frequencies, and temporal patterns will be formed for each of them. The next step will be cross-comp among temporally adjacent patterns of different frequencies, similar to comp_slice_.
> Yes, we will use the human audible spectrum, but we don't know what that set of recurring frequencies is, within the spectrum.
> Did you see my last top comment?

So based on this diagram, each recurring frequency you mentioned should be an unique wave in time domain right? In this diagram, there are 3 recurring frequencies?

![image](https://user-images.githubusercontent.com/52521979/114703744-42e63c80-9d58-11eb-8623-861fc804c9e5.png)

In an actual scenario, those waves should be more complex, it may having very much more of those recurring frequencies, and what we are trying to do is group those recurring frequency (same with grouping pixels to pattern)? However, this should be now do-able by converting them to the frequency domain by using FFT right?

Also in current context, the 'spectrum' is the same with spectrum in frequency domain？ Wouldn't it should be defined by the detect-able recurring frequencies and amplitude of the waves?

**Reply by boris-kz** [reply]
Spectrum may mean all detectable frequencies, or a set of recurring (actually detected) frequencies. I use it for the former, the later we can call something like "channels"?

> what we are trying to do is group those recurring frequency (same with grouping pixels to pattern)?

Yes, I think this grouping should be done by cross-comparing recurring frequencies, which would be obtained by incremental-range cross-comp of above-average intensity that I mentioned? Could you look into details of how DFT does it?

**Reply by kwcckw** [reply]
> 
> 
> Spectrum may mean all detectable frequencies, or a set of recurring (actually detected) frequencies. I use it for the former, the later we can call something like "channels"?

Yea, so similar with image, grayscale is having 1 channel, while rgb(colour image) is having 3 channels. But in this case, we should always having 1 channel / spectrum? Or number of channel should be defined by us or the data itself?

> 
> > what we are trying to do is group those recurring frequency (same with grouping pixels to pattern)?
> 
> Yes, I think this grouping should be done by cross-comparing recurring frequencies, which would be obtained by incremental-range cross-comp of above-average intensity that I mentioned? Could you look into details of how DFT does it?

Yes, it seems feasible with range comparison, since they should be forming certain pattern regularly if they are having the same frequency. I will look into the DFT and FFT and check again, since it's been sometime since the last time i work on frequency domain, need to read through them again.

**Reply by boris-kz** [reply]
Raw audio is PCM, only one channel: https://en.wikipedia.org/wiki/Pulse-code_modulation.
Active channels are sample-specific and need to be discovered.

**Reply by kwcckw** [reply]
I think this link is quite informative on how to derive the DFT equations:
https://www.allaboutcircuits.com/technical-articles/an-introduction-to-the-discrete-fourier-transform/

I'm still trying to understand those equations and relate it to our case, will update again.

**Reply by boris-kz** [reply]
Thanks Chee!

**Reply by aqibmumtaz** [reply]
> @aqibmumtaz, yes intensity is amplitude here, it means the same thing, I want to use common terms for both images and audio.
> 
> > it means when we convert time to frequency domain for one time instant we have multiple frequencies?
> 
> Yes, is that how it works in DFT? I need to get into details of FT, does it work like above-minimal-amplitude cross-comparison at incremental distance, as I proposed in the last comment?
> 
> > you are talking about time-frequency overlapped signals?
> 
> I mistakenly assumed that frequency is a parameter of input in raw audio, kind of like colors in images. It turned out that we have to compute component frequencies via some kind FT first.
> 
> Then we form time-domain patterns of intensity for each component frequency, with current line_patterns. And then we cross-compare time-overlapped patterns between frequencies, similar to comp_slice_ module in 2D alg. So, time domain will be a primary dimension, in which we define Ps, and frequency domain is secondary, in which we define time-frequency overlapped PPs.
> 
> Does it make sense?

Yeah frequency components can be obtained from FT or in audio signal more likely from FFT. then we will have amplitude versus frequency. I have got the idea of Time-frequency overlapped signals and to find pattern between them but I have went through 2D alg yet cant really get the last thing you said . what I have understood is from time domain we will form patterns Ps as a amplitude input and then the PPs from time-frequency overlapped domain?

**Reply by boris-kz** [reply]
Yes, but that time-frequency overlap is with Ps, not primary inputs. 
I guess we need direct FFT on primary input, then reverse FFT on each resulting frequency separately.
Then line_patterns will process in time domain on each frequency channel, forming Ps.
Then we align resulting Ps between channels in time, and compare time-overlapping Ps between frequencies, forming PPs.
But in that comp_P, I don't know how important is proximity in frequency domain: do we compare with incremental difference between frequencies, or between all frequencies at once?

**Reply by kwcckw** [reply]
> 
> 
> Thanks Chee!

I think this video is very helpful to visualize on the fourier transform:
https://www.youtube.com/watch?v=spUNpyF58BY&list=RDCMUCYO_jab_esuFRV4b17AJtAw&index=1

From the video, i think what we can do is finding the repeating patterns across the signal, and we need an ave for the frequency of repeating to add that as recurring frequency. Here's my suggestion:

![image](https://user-images.githubusercontent.com/52521979/115020577-5c6cbd00-9eed-11eb-88fb-10591645c8bd.png)

**Reply by boris-kz** [reply]
Right, but we also need to specify that match a period (L+neg_L) has an extra weight in crest parameters. 
Closer crest parameters: this would require using indirect match: ave - abs diff, vs. match = min(P1, P2).
Normally, we would use direct match on this level, but it seems that we have to go back to indirect match after forming params on a new level of composition (Ps vs. derts in this case)?

**Reply by kwcckw** [reply]
Yes, L + neg_L should be the T of the recurring frequency, so it is one of the most important parameter with more weight.

Regarding the match , so mostly we will form new match related parameters from both Ps parameters. But what would be the reason to use indirect match instead of direct match? Is it indirect match for derived parameters?

**Reply by kwcckw** [reply]
Also the suggestion above should be done in time domain, without translating it into frequency domain, since basically what we want to do is similar with Fourier transform but in time domain.

**Reply by aqibmumtaz** [reply]
> Yes, but that time-frequency overlap is with Ps, not primary inputs.
> I guess we need direct FFT on primary input, then reverse FFT on each resulting frequency separately.
> Then line_patterns will process in time domain on each frequency channel, forming Ps.
> Then we align resulting Ps between channels in time, and compare time-overlapping Ps between frequencies, forming PPs.
> But in that comp_P, I don't know how important is proximity in frequency domain: do we compare with incremental difference between frequencies, or between all frequencies at once?

Alright
Ps will be find between each frequency channels in time domain and then PPs form overlapping Ps. In frequency domain I think comparison will be between all frequency at once because in frequency domain we can't define or have wavelength which you are defining easily in time domain.

**Reply by boris-kz** [reply]
Proximity in frequency domain still matters: the mechanism for generating sounds (such as vocal cords) is far more likely to drift or switch into neighboring frequencies. Also, voices of different people differ by a relatively small shift in frequency.
So, while comparison range will be greater in frequency domain, we still have to account for the distance.

**Reply by boris-kz** [reply]
> Yes, L + neg_L should be the T of the recurring frequency, so it is one of the most important parameter with more weight.
> 
> Regarding the match , so mostly we will form new match related parameters from both Ps parameters. But what would be the reason to use indirect match instead of direct match? Is it indirect match for derived parameters?

Indirect match is used for input variables where there is low correlation between magnitude and predictive value (inertia).
That's not the case with sound, amplitude is proportional to the effort needed to generate it.
But min is basically a co-detection, it doesn't depend difference. So, it's good for merging above-average patterns, but doesn't look like a similarity measure on it's own.
We may need to use projected match: min - abs diff / 2. Need to think about it.

**Reply by aqibmumtaz** [reply]
> Proximity in frequency domain still matters: the mechanism for generating sounds (such as vocal cords) is far more likely to drift or switch into neighboring frequencies. Also, voices of different people differ by a relatively small shift in frequency.
> So, while comparison range will be greater in frequency domain, we still have to account for the distance.

and how to calculate comparison range in frequency domain? just like what the range-comp in line_patterns?

**Reply by boris-kz** [reply]
It's more like comp_slice_ in 2D alg, blob slice is just a more intuitive term for P here.
But at this point it only compares vertically adjacent Ps. So yes, like in range_comp we can use incremental range, but it will probably be incremental in the number of consecutive cross-compared frequencies, not in the difference in frequency. This would be selective per PP, defined by prior-range cross-comp.

**Comment by boris-kz** [comment]
I just checked for the first time, uncompressed audio is PCM, which doesn't represent frequency, only amplitude.
To find locally salient frequencies, we need either something like Fourier transform as a shortcut, or algorithm itself should discover them, which is more consistent. That can probably be done by a version of range_comp, except it will be storing individual matches, vs. summing them in a kernel. I was thinking about this version before, it makes sense for inputs that only match at a specific range. Then we compare these individual matches of the same input at different distances, to find the distance at which they frequently match. That distance is a wavelength. Not even sure this process is much different from FT?

**Reply by kwcckw** [reply]
> 
> 
> I just checked for the first time, uncompressed audio is PCM, which doesn't represent frequency, only amplitude.
> To find locally salient frequencies, we need either something like Fourier transform as a shortcut, or algorithm itself should discover them, which is more consistent. That can probably be done by a version of range_comp, except it will be storing individual matches, vs. summing them in a kernel. I was thinking about this version before, it makes sense for inputs that only match at a specific range. Then we compare these individual matches of the same input at different distances, to find the distance at which they frequently match. That distance is a wavelength. Not even sure this process is much different from FT?

I'm not really familiar in Fourier transform but this would looks partially similar. Since it would change the way we represent the data for the purpose of finding pattern within it.

**Reply by boris-kz** [reply]
Actually, that frequency discovery won't be by range_comp in line_patterns, it will be done in line_PPs. That's because it's much easier to justify incremental-distance cross-comp of wave crests, vs. individual pixels. Match in sound is defined directly: as min intensity, thus positive patterns are spans of above-average intensity, which will be around those wave crests. This cross-comp P will also be done on negative Ps (spans between crests) and complemented +P,-P. That complemented span is a wavelength,  the span over which it cross-matches is a PP: the span of recurring frequency we are looking for.

**Reply by kwcckw** [reply]
> 
> 
> Actually, that frequency discovery won't be by range_comp in line_patterns, it will be done in line_PPs. That's because it's much easier to justify incremental-distance cross-comp of wave crests, vs. individual pixels. 

Yea, this is more reasonable since it's similar with image , we didn't perform recognition/detection with pixels but with features instead. So in this case, we are looking for similar recurring pattern within the frequency with some detected pattern, instead of single data point pattern. So incremental-distance cross-comp of wave crests should be better than individual data point.

> Match in sound is defined directly: as min intensity, thus positive patterns are spans of above-average intensity, which will be around those wave crests. This cross-comp P will also be done on negative Ps (spans between crests) and complemented +P,-P. That complemented span is a wavelength, the span over which it cross-matches is a PP: the span of recurring frequency we are looking for.

From my understanding, this should be the case but i'm not so sure about the PP, is that correct? Or PP should be all +P and -P pair?

![image](https://user-images.githubusercontent.com/52521979/114802134-becf9b80-9dcf-11eb-948a-c7bf18bd2a7f.png)

**Comment by boris-kz** [comment]
PP is a span of matching Ps, as P is a span of matching pixels. That match is computed for each P parameter, including L and neg_L.
Matching L + neg_L means PP represents a span of recurring frequency.

**Reply by kwcckw** [reply]
ah i see,  and in current case, other than L, what would be the other potential matching parameter? Such as amplitude, mA? time mT? and right now all of these are applicable to time domain only right?

**Reply by boris-kz** [reply]
Basically, this should be line_PPs as we have it now, I see no reason to modify it yet. Again, I want to keep the the same terms across images and audio as much as possible, so it's still intensity, match, and and L as we have in code. There is no need for mT, match of coordinate is just inverse distance, which is expressed by L + neg_L. But we need to add this distance to comp_P, and cross-comp of distance (to find recurring wavelength) will probably be a incremental-derivation version of comp_P. It will be similar to deriv__comp in line_patterns, but for distances vs. differences, and inside positive PPm only.

**Reply by boris-kz** [reply]
So, the process in FFT is very different, it's based on integration vs. comparison. There should be salient frequency selection (>ave) after integration too, but I haven't seen it yet? This processes is specialized vs. our general, so it's obviously a lot faster. We  need to explain how line_PPs can find salient frequencies, but use FFT to actually do that. Then run time-domain line_patterns within each frequency. Next is probably a version of comp_slice_ (without segment by direction), comparing time-overlapping Ps across adjacent frequencies, so it will be 2D. What do you think @aqibmumtaz?

**Reply by aqibmumtaz** [reply]
What I have understood is we have to go step vise like this
1. Calculate Ps in time domain after applying FFT on the primary input(that is applying direct fft on the audio signal), then applying ifft on each frequency and then calculating Ps for each frequency channel by applying 1-D line-pattern algo.

**Comment by aqibmumtaz** [comment]
and I am assuming our all the discussion is for mono audio file....Am I right @boris-kz ?
As for mono the extracted data array would be 1-D and for stereo we would have 2-D data array?

**Reply by boris-kz** [reply]
Yes, mono, The 2D I mentioned is frequency.

**Reply by aqibmumtaz** [reply]
frequency is 2-D or time-frequency overlapped signal will be 2-D?
because i didnt get 2-D frequency point

**Reply by boris-kz** [reply]
Sorry, I meant that frequency is the 2nd dimension, time is the 1st one.
